proxy/request.go

```diff
@@ -118,32 +118,120 @@ func (r *request) OnClose(_ error) {
 	} else {
 		if !r.done {
 			r.done = true
-			r.send(&message.Unavailable{ErrorMessage: "No more hosts available (cluster connection closed and request is not idempotent)"})
+			r.send(&message.ServerError{ErrorMessage: "Proxy is unable to retry non-idempotent query after connection to backend cluster closed"})
 		}
 	}
 }
 
 func (r *request) OnResult(raw *frame.RawFrame) {
 	r.mu.Lock()
+	defer r.mu.Unlock()
 	if !r.done {
-		r.done = true
-		if r.raw.Header.OpCode == primitive.OpCodePrepare && raw.Header.OpCode == primitive.OpCodeResult { // Prepared result
-			frm, err := codec.ConvertFromRawFrame(raw)
-			if err != nil {
-				r.client.proxy.logger.Error("error attempting to decode prepared result message")
-			} else if _, ok := frm.Body.Message.(*message.PreparedResult); !ok { // TODO: Use prepared type data to disambiguate idempotency
-				r.client.proxy.logger.Error("expected prepared result message, but got something else")
-			} else {
-				idempotent, err := parser.IsQueryIdempotent(r.query)
+		if raw.Header.OpCode != primitive.OpCodeError ||
+			!r.handleErrorResult(raw) { // If the error result is retried then we don't send back this response
+			if r.raw.Header.OpCode == primitive.OpCodePrepare && raw.Header.OpCode == primitive.OpCodeResult { // Prepared result
+				frm, err := codec.ConvertFromRawFrame(raw)
 				if err != nil {
-					r.client.proxy.logger.Error("error parsing query for idempotence", zap.Error(err))
+					r.client.proxy.logger.Error("error attempting to decode prepared result message")
+				} else if _, ok := frm.Body.Message.(*message.PreparedResult); !ok { // TODO: Use prepared type data to disambiguate idempotency
+					r.client.proxy.logger.Error("expected prepared result message, but got something else")
 				} else {
-					// TODO: Make sure this hash matches server-side impl.
-					r.client.preparedIdempotence.Store(md5.Sum([]byte(r.keyspace+r.query)), idempotent)
+					idempotent, err := parser.IsQueryIdempotent(r.query)
+					if err != nil {
+						r.client.proxy.logger.Error("error parsing query for idempotence", zap.Error(err))
+					} else {
+						// TODO: Make sure this hash matches server-side impl.
+						r.client.preparedIdempotence.Store(md5.Sum([]byte(r.keyspace+r.query)), idempotent)
```

Comment:
This doesn't match in the case of Astra prepared IDs, Idk if it causes issues or not

Reply:
Oh no! I need to dig into that then.

Reply:
I'm able to avoid this by using the prepared ID from the response (which is what I should have been doing anyway):

```go
} else if result, ok := frm.Body.Message.(*message.PreparedResult); ok {
   r.client.preparedIdempotence.Store(preparedIdKey(result.PreparedQueryId), idempotent)
} else {
  // ...
```

Reply:
Yeah that's what ZDM is doing as well.

Reply:
Thanks for questioning this.